Edit "Queries" chapter (#1301)
rust-lang/rustc-dev-guide: src/query.md

# Queries: demand-driven compilation

<!-- toc -->

As described in [the high-level overview of the compiler][hl], the Rust compiler
is still (as of <!-- date: 2021-07 --> July 2021) transitioning from a
traditional "pass-based" setup to a "demand-driven" system. The compiler query

> system is the key to rustc's demand-driven organization.

Reviewer: I think it might still make sense to have this in bold.

Author: I generally don't like using bold face for emphasis in regular prose. Let me try something else here...

Author: Actually, I'm thinking of leaving this as-is, if that's alright. 😇

Reviewer: Eh, not a big deal either way. I slightly lean toward bold in this case, but it's fine that you left it out.

The idea is pretty simple. Instead of entirely independent passes

> (parsing, type-checking, etc.), various smaller *queries*

Reviewer: ```suggestion
(parsing, type-checking, etc.), a set of function-like *queries*
```

compute information about the input source. For example,

> there is a query called `type_of(def_id)` that, given the [`DefId`] of

Reviewer: ```suggestion
there is a query called `type_of` that, given the [`DefId`] of
```

some item, will compute the type of that item and return it to you.

> [`DefId`]: appendix/glossary.md#def-id

Reviewer: I think we should link to the docs here; that's what we usually do.

[hl]: ./compiler-src.md

Query execution is *memoized*. The first time you invoke a
query, it will go do the computation, but the next time, the result is
returned from a hashtable. Moreover, query execution fits nicely into

> *incremental computation*; the idea is roughly that, when you do a

Reviewer: ```suggestion
*incremental computation*; the idea is roughly that, when you invoke a
```

query, the result *may* be returned to you by loading stored data
from disk.[^1]

Eventually, we want the entire compiler
control-flow to be query driven. There will effectively be one
top-level query (`compile`) that will run compilation on a crate; this
will in turn demand information about that crate, starting from the
*end*.  For example:

- The `compile` query might demand to get a list of codegen-units
  (i.e. modules that need to be compiled by LLVM).
- But computing the list of codegen-units would invoke some subquery
  that returns the list of all modules defined in the Rust source.
- That query in turn would invoke something asking for the HIR.
- This keeps going further and further back until we wind up doing the
  actual parsing.

> Although this vision is not fully realized, large sections of

Reviewer: ```suggestion
Although this vision is not fully realized, large sections of the
```

> compiler (for example, generating [MIR](mir/)) currently work exactly like this.

Reviewer: ```suggestion
compiler (for example, generating [MIR](./mir/)) currently work exactly like this.
```

### Invoking queries

> Invoking a query is simple. The [`TyTcx`] ("type context") struct offers a method

Reviewer: You'll need to update the link ref name too.

```suggestion
Invoking a query is simple. The [`TyCtxt`] ("type context") struct offers a method
```

[rustc_metadata]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/index.html
[wasm_import_module_map]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/back/symbol_export/fn.wasm_import_module_map.html

### Adding a new query

> Suppose you want to add a new query: how do you do so?

Reviewer: ```suggestion
How do you add a new query?
```

```

Queries are grouped into categories (`Other`, `Codegen`, `TypeChecking`, etc.).
Each group contains one or more queries.

A query definition has the following form:

```rust,ignore

> query type_of(key: DefId) -> Ty<'tcx> { ... }

Reviewer: The `^^` for "query keyword" should point to the whole word.

#### Query structs and descriptions

For each query, the `rustc_queries` macro will generate a "query struct"

> named after the query. This struct is a kind of a place-holder

Reviewer: ```suggestion
named after the query. This struct is a kind of placeholder
```

[GitHub issue #42633]: https://github.com/rust-lang/rust/issues/42633
[Incremental Compilation Beta]: https://internals.rust-lang.org/t/incremental-compilation-beta/4721
[Incremental Compilation Announcement]: https://blog.rust-lang.org/2016/09/08/incremental.html

[^1]: The ["Incremental Compilation in Detail](queries/incremental-compilation-in-detail.md) chapter gives a more
in-depth description of what queries are and how they work.

> If you intend to write a query of your own, this is a good read.

Reviewer: Can you move the footnote to near the reference site and rename `^1` to something explanatory? And the same for the second footnote.
